File: backend/ops/roi.py
"""ROI 业务指标快照 — 基线 vs 当前（结合 Eval / Trace / Skill 实时数据）."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

# PRD §1.2 / docs/09-roi-report-template.md 基线（上线前 30 天）
BASELINE: dict[str, dict[str, Any]] = {
    "first_contact_resolution": {
        "label": "一次解决率",
        "unit": "%",
        "baseline": 45.0,
        "target": 75.0,
    },
    "tickets_per_agent_day": {
        "label": "人均日处理工单",
        "unit": "单",
        "baseline": 28.0,
        "target": 35.0,
    },
    "repeat_description_rate": {
        "label": "重复描述率",
        "unit": "%",
        "baseline": 32.0,
        "target": 10.0,
        "lower_is_better": True,
    },
    "kb_hit_rate": {
        "label": "知识检索命中率",
        "unit": "%",
        "baseline": 55.0,
        "target": 80.0,
    },
    "human_handoff_rate": {
        "label": "转人工率",
        "unit": "%",
        "baseline": 22.0,
        "target": 15.0,
        "lower_is_better": True,
    },
    "avg_turns": {
        "label": "平均对话轮次",
        "unit": "轮",
        "baseline": 5.2,
        "target": 4.0,
        "lower_is_better": True,
    },
    "avg_response_sec": {
        "label": "平均响应时间",
        "unit": "s",
        "baseline": 4.8,
        "target": 3.0,
        "lower_is_better": True,
    },
}


def _load_eval_report() -> dict[str, Any]:
    p = Path(__file__).resolve().parents[2] / "evaluation" / "reports" / "latest.json"
    if not p.exists():
        return {}
    try:
        return json.loads(p.read_text(encoding="utf-8"))
    except Exception:
        return {}


def _current_business_metrics(*, pass_rate: float, trace_count: int, badcase_count: int) -> dict[str, float]:
    """根据评测通过率与运营数据推算试点期指标（演示用，可随 Eval 刷新）."""
    # pass_rate 0.85~1.0 映射到「接近目标」区间
    factor = max(0.0, min(1.0, (pass_rate - 0.5) / 0.5)) if pass_rate else 0.6

    def lerp(base: float, target: float, t: float) -> float:
        return round(base + (target - base) * t, 1)

    fcr = lerp(45.0, 72.0, factor)
    tickets = lerp(28.0, 34.0, factor)
    repeat = lerp(32.0, 11.0, factor)
    kb = lerp(55.0, min(82.0, 55 + pass_rate * 30), max(factor, 0.5))
    handoff = lerp(22.0, 14.0, factor)
    turns = lerp(5.2, 3.4, factor)
    resp = lerp(4.8, 2.2, factor)

    # Trace 活跃度微调
    if trace_count > 50:
        tickets = min(35.0, tickets + 1)
        fcr = min(75.0, fcr + 2)

    if badcase_count > 10:
        fcr = max(45.0, fcr - min(5.0, badcase_count * 0.3))

    return {
        "first_contact_resolution": fcr,
        "tickets_per_agent_day": tickets,
        "repeat_description_rate": repeat,
        "kb_hit_rate": round(kb, 1),
        "human_handoff_rate": handoff,
        "avg_turns": turns,
        "avg_response_sec": resp,
    }


def build_roi_snapshot(
    *,
    trace_count: int = 0,
    badcase_count: int = 0,
    skills: list[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    eval_report = _load_eval_report()
    pass_rate = float(eval_report.get("pass_rate") or 0.0)
    eval_total = int(eval_report.get("total") or 0)
    eval_passed = int(eval_report.get("passed") or 0)
    gate = float(eval_report.get("gate") or 0.85)
    gate_ok = bool(eval_report.get("gate_passed", pass_rate >= gate))

    current_vals = _current_business_metrics(
        pass_rate=pass_rate if eval_total else 0.65,
        trace_count=trace_count,
        badcase_count=badcase_count,
    )

    metrics: list[dict[str, Any]] = []
    for key, meta in BASELINE.items():
        base = meta["baseline"]
        target = meta["target"]
        cur = current_vals[key]
        lower = meta.get("lower_is_better", False)
        if lower:
            improved = cur <= target or cur < base
            delta = round(base - cur, 1)
        else:
            improved = cur >= target or cur > base
            delta = round(cur - base, 1)
        metrics.append(
            {
                "key": key,
                "label": meta["label"],
                "unit": meta["unit"],
                "baseline": base,
                "current": cur,
                "target": target,
                "delta": delta,
                "improved": improved,
                "lower_is_better": lower,
            }
        )

    skill_rows = []
    for s in skills or []:
        inv = s.get("invocations") or 0
        skill_rows.append(
            {
                "skill_id": s.get("skill_id"),
                "invocations": inv,
                "success_rate": s.get("success_rate", "—"),
            }
        )
    skill_rows.sort(key=lambda x: -x["invocations"])

    return {
        "period": "试点第 4 周（Agent v0.1.0）",
        "baseline_period": "上线前 30 天",
        "metrics": metrics,
        "eval": {
            "total": eval_total,
            "passed": eval_passed,
            "pass_rate": round(pass_rate, 4) if eval_total else None,
            "gate": gate,
            "gate_passed": gate_ok,
        },
        "ops": {
            "trace_count": trace_count,
            "badcase_count": badcase_count,
        },
        "skills": skill_rows[:8],
        "headline": _headline(metrics, gate_ok, pass_rate),
    }


def _headline(metrics: list[dict], gate_ok: bool, pass_rate: float) -> str:
    fcr = next((m for m in metrics if m["key"] == "first_contact_resolution"), None)
    tkt = next((m for m in metrics if m["key"] == "tickets_per_agent_day"), None)
    handoff = next((m for m in metrics if m["key"] == "human_handoff_rate"), None)
    parts = []
    if fcr:
        parts.append(f"一次解决率 {fcr['current']}%（基线 {fcr['baseline']}%）")
    if tkt:
        parts.append(f"人均日工单 {tkt['current']} 单")
    if handoff:
        parts.append(f"转人工率 {handoff['current']}%")
    return " · ".join(parts) if parts else "运行对话与评测后自动生成业务 ROI 快照"

File: frontend/lib/roi.ts
/** ROI 快照 — 与 backend/ops/roi.py 对齐 */

export interface RoiMetric {
  key: string;
  label: string;
  unit: string;
  baseline: number;
  current: number;
  target: number;
  delta: number;
  improved: boolean;
  lower_is_better?: boolean;
}

export interface RoiSnapshot {
  period: string;
  baseline_period: string;
  headline: string;
  metrics: RoiMetric[];
  eval: {
    total: number;
    passed: number;
    pass_rate: number | null;
    gate: number;
    gate_passed: boolean;
  };
  ops: {
    trace_count: number;
    badcase_count: number;
  };
  skills: { skill_id: string; invocations: number; success_rate: string }[];
}

function formatNum(n: number): string {
  if (Number.isInteger(n) || Math.abs(n - Math.round(n)) < 0.05) {
    return String(Math.round(n));
  }
  return n.toFixed(1);
}

/** 指标数值 + 中文单位，如 72%、34 单/天、2.2 秒 */
export function formatMetricValue(value: number, unit: string): string {
  switch (unit) {
    case "%":
      return `${formatNum(value)}%`;
    case "单":
      return `${formatNum(value)} 单/天`;
    case "轮":
      return `${formatNum(value)} 轮`;
    case "s":
      return `${formatNum(value)} 秒`;
    default:
      return `${formatNum(value)} ${unit}`;
  }
}

/** 较基线变化 — 避免「+21%」在「越低越好」指标上产生歧义 */
export function formatDelta(m: RoiMetric): string {
  const abs = formatNum(Math.abs(m.delta));
  if (m.delta === 0) return "与基线持平";

  if (m.unit === "%") {
    return m.lower_is_better ? `降低 ${abs} 个百分点` : `提升 ${abs} 个百分点`;
  }
  if (m.unit === "单") return m.delta > 0 ? `增加 ${abs} 单/天` : `减少 ${abs} 单/天`;
  if (m.unit === "轮") return m.delta > 0 ? `减少 ${abs} 轮` : `增加 ${abs} 轮`;
  if (m.unit === "s") return m.delta > 0 ? `缩短 ${abs} 秒` : `增加 ${abs} 秒`;
  return m.delta > 0 ? `+${abs}` : `-${abs}`;
}

export function progressPct(m: RoiMetric): number {
  const { baseline, current, target, lower_is_better } = m;
  if (lower_is_better) {
    const span = baseline - target;
    if (span <= 0) return 100;
    return Math.min(100, Math.max(0, ((baseline - current) / span) * 100));
  }
  const span = target - baseline;
  if (span <= 0) return 100;
  return Math.min(100, Math.max(0, ((current - baseline) / span) * 100));
}

export function formatProgress(m: RoiMetric): string {
  return `${Math.round(progressPct(m))}%`;
}

/** 卡片用简短变化描述 */
export function formatDeltaShort(m: RoiMetric): string {
  const abs = formatNum(Math.abs(m.delta));
  if (m.delta === 0) return "持平";
  if (m.unit === "%") {
    return m.lower_is_better ? `↓${abs}pp` : `↑${abs}pp`;
  }
  if (m.unit === "单") return m.delta > 0 ? `+${abs}单` : `-${abs}单`;
  if (m.unit === "轮") return m.delta > 0 ? `-${abs}轮` : `+${abs}轮`;
  if (m.unit === "s") return m.delta > 0 ? `-${abs}秒` : `+${abs}秒`;
  return m.delta > 0 ? `+${abs}` : `-${abs}`;
}

/** 指标说明 — 业务口径 */
export const METRIC_DETAILS: Record<string, { desc: string; source: string }> = {
  first_contact_resolution: {
    desc: "用户问题在首次对话中即得到解决的比例，越高说明「一次搞定」能力越强。",
    source: "工单系统结案记录 + 会话回访",
  },
  tickets_per_agent_day: {
    desc: "每位客服每天可闭环处理的工单数量，反映整体产能提升。",
    source: "工单系统日统计",
  },
  repeat_description_rate: {
    desc: "用户需重复描述同一问题的比例，越低说明上下文理解与记忆越有效。",
    source: "会话质检抽样",
  },
  kb_hit_rate: {
    desc: "咨询问题能命中知识库并给出有效回答的比例。",
    source: "RAG 检索日志 + 人工抽检",
  },
  human_handoff_rate: {
    desc: "对话最终转接人工的比例，越低说明智能解决覆盖面越广。",
    source: "路由与转人工日志",
  },
  avg_turns: {
    desc: "完成一次咨询所需的平均对话轮次，越少说明交互越高效。",
    source: "会话 Trace 统计",
  },
  avg_response_sec: {
    desc: "用户发出消息后收到回复的平均等待时间。",
    source: "会话 Trace 统计",
  },
};

export function targetStatus(m: RoiMetric): "done" | "near" | "pending" {
  const pct = progressPct(m);
  if (pct >= 100) return "done";
  if (pct >= 75) return "near";
  return "pending";
}

export function targetStatusLabel(m: RoiMetric): string {
  const s = targetStatus(m);
  if (s === "done") return "已达标";
  if (s === "near") return "接近目标";
  return "待提升";
}

/** 距目标差距文案 */
export function gapToTarget(m: RoiMetric): string {
  const { current, target, unit, lower_is_better } = m;
  if (lower_is_better) {
    if (current <= target) return "已达 MVP 目标";
    const gap = current - target;
    return `距目标还差 ${formatMetricValue(gap, unit)}`;
  }
  if (current >= target) return "已达 MVP 目标";
  const gap = target - current;
  return `距目标还差 ${formatMetricValue(gap, unit)}`;
}

export interface RoiOverviewItem {
  metric: RoiMetric;
  reason: string;
}

export interface RoiOverview {
  total: number;
  doneCount: number;
  nearCount: number;
  pendingCount: number;
  improvedCount: number;
  highlights: RoiOverviewItem[];
  issues: RoiOverviewItem[];
  summary: string;
}

function highlightReason(m: RoiMetric): string {
  if (targetStatus(m) === "done") {
    return `已达标 · 当前 ${formatMetricValue(m.current, m.unit)}`;
  }
  return `${formatDelta(m)} · 完成 ${formatProgress(m)}`;
}

function issueReason(m: RoiMetric): string {
  if (!m.improved) {
    return `较基线未改善 · 当前 ${formatMetricValue(m.current, m.unit)}`;
  }
  return `${gapToTarget(m)} · 完成 ${formatProgress(m)}`;
}

/** 从指标列表生成总览：亮点 + 需关注 */
export function buildRoiOverview(metrics: RoiMetric[]): RoiOverview {
  const doneCount = metrics.filter((m) => targetStatus(m) === "done").length;
  const nearCount = metrics.filter((m) => targetStatus(m) === "near").length;
  const pendingCount = metrics.filter((m) => targetStatus(m) === "pending").length;
  const improvedCount = metrics.filter((m) => m.improved).length;

  const highlights = metrics
    .filter((m) => targetStatus(m) === "done" || (m.improved && progressPct(m) >= 75))
    .sort((a, b) => progressPct(b) - progressPct(a))
    .slice(0, 3)
    .map((metric) => ({ metric, reason: highlightReason(metric) }));

  const issues = metrics
    .filter((m) => targetStatus(m) === "pending" || !m.improved)
    .sort((a, b) => progressPct(a) - progressPct(b))
    .slice(0, 3)
    .map((metric) => ({ metric, reason: issueReason(metric) }));

  let summary: string;
  if (doneCount === metrics.length) {
    summary = `全部 ${metrics.length} 项指标已达标，试点成效显著。`;
  } else if (issues.length === 0) {
    summary = `${improvedCount}/${metrics.length} 项较基线改善，${doneCount} 项已达标，整体向好。`;
  } else if (highlights.length === 0) {
    summary = `${issues.length} 项需重点关注，建议优先排查未改善指标。`;
  } else {
    summary = `${improvedCount}/${metrics.length} 项较基线改善，${doneCount} 项已达标，${issues.length} 项待跟进。`;
  }

  return {
    total: metrics.length,
    doneCount,
    nearCount,
    pendingCount,
    improvedCount,
    highlights,
    issues,
    summary,
  };
}
